python/ray/_private/gcs_pubsub.py

```diff
@@ -94,108 +122,181 @@ def publish_error(self, key_id: bytes, error_info: ErrorTableData) -> None:
-                        # and return from polling.
-                        if self._close.is_set():
-                            fut.cancel()
-                            return None, None
-                        # GRPC has not replied, continue waiting.
+            fut = self._stub.GcsSubscriberPoll.future(
+                self._poll_request(), timeout=timeout)
+            # Wait for result to become available, or cancel if the
+            # subscriber has closed.
+            while True:
+                # Unlock before waiting for reply.
+                self._lock.release()
+                try:
+                    # Use 1s timeout to check for subscriber closing
+                    # periodically.
+                    fut.result(timeout=1)
+                    break
+                except grpc.FutureTimeoutError:
+                    # Subscriber has closed. Cancel inflight request and
+                    # return from polling.
+                    if self._close.is_set():
+                        fut.cancel()
+                        return
+                    # GRPC has not replied, continue waiting.
+                    continue
+                except grpc.RpcError as e:
+                    # Choose to not raise deadline exceeded errors to the
+                    # caller. Instead return None. This can be revisited later.
+                    if e.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
+                        return
+                    raise
+                finally:
+                    # Make sure lock is held if exiting from the loop.
+                    self._lock.acquire()
+
+            if fut.done():
+                for msg in fut.result().pub_messages:
+                    if msg.channel_type != self._channel:
+                        logger.warn(
+                            f"Ignoring message from unsubscribed channel {msg}"
+                        )
                         continue
-                    except Exception:
-                        # GRPC error, including deadline exceeded.
-                        raise
-                if fut.done():
-                    for msg in fut.result().pub_messages:
-                        self._errors.append((msg.key_id,
-                                             msg.error_info_message))
-
-            if len(self._errors) == 0:
-                return None, None
-            return self._errors.popleft()
+                    self._queue.append(msg)
 
     def close(self) -> None:
-        """Closes the subscriber and its active subscriptions."""
+        """Closes the subscriber and its active subscription."""
+
         # Mark close to terminate inflight polling and prevent future requests.
         self._close.set()
+        req = self._unsubscribe_request(channels=[self._channel])
+        try:
+            self._stub.GcsSubscriberCommandBatch(req, timeout=30)
+        except Exception:
```

Comment:
Why do we need this call?

Reply:
This removes the subscription state from publisher in GCS. Currently we still rely on publisher GC to remove other subscriber state, so the benefit may not be large yet. But I think it is a better practice to signal termination of the subscriber here. In future publisher can remove all related data based on this call, instead of waiting for 30~60s to GC all subscriber state. I will use a shorter deadline here to avoid slowing down close() too much.